packages/flutter_test/lib/src/finders.dart

```diff
@@ -520,121 +907,225 @@ abstract class Finder {
   /// target element where the hit test is performed.
-  Finder hitTestable({ Alignment at = Alignment.center }) => _HitTestableFinder(this, at);
+  WidgetFinder hitTestable({ Alignment at = Alignment.center }) => _HitTestableWidgetFinder(this, at);
+}
+
+/// A base class for creating finders that search the semantics tree.
+abstract class SemanticsFinder extends FinderBase<SemanticsNode> {
+  /// Creates a new [SemanticsFinder] that will search starting at the given
+  /// [root].
+  SemanticsFinder(this.root);
+
+  /// The root of the semantics tree that this finder will search.
+  final SemanticsNode root;
 
   @override
-  String toString() {
-    final String additional = skipOffstage ? ' (ignoring offstage widgets)' : '';
-    final List<Element> widgets = evaluate().toList();
-    final int count = widgets.length;
-    if (count == 0) {
-      return 'zero widgets with $description$additional';
-    }
-    if (count == 1) {
-      return 'exactly one widget with $description$additional: ${widgets.single}';
-    }
-    if (count < 4) {
-      return '$count widgets with $description$additional: $widgets';
-    }
-    return '$count widgets with $description$additional: ${widgets[0]}, ${widgets[1]}, ${widgets[2]}, ...';
+  Iterable<SemanticsNode> get allCandidates {
+    return collectAllSemanticsNodesFrom(root);
   }
 }
 
-/// Applies additional filtering against a [parent] [Finder].
-abstract class ChainedFinder extends Finder {
-  /// Create a Finder chained against the candidates of another [Finder].
-  ChainedFinder(this.parent);
+/// A mixin that applies additional filtering to the results of a parent [Finder].
+ mixin ChainedFinderMixin<CandidateType> on FinderBase<CandidateType> {
 
-  /// Another [Finder] that will run first.
-  final Finder parent;
+  /// Another finder whose results will be further filtered.
+  FinderBase<CandidateType> get parent;
 
   /// Return another [Iterable] when given an [Iterable] of candidates from a
-  /// parent [Finder].
+  /// parent [FinderBase].
   ///
-  /// This is the method to implement when subclassing [ChainedFinder].
-  Iterable<Element> filter(Iterable<Element> parentCandidates);
+  /// This is the method to implement when mixing in [ChainedFinderMixin] or
+  /// extending [ChainedWidgetFinder].
+  Iterable<CandidateType> filter(Iterable<CandidateType> parentCandidates);
 
   @override
-  Iterable<Element> apply(Iterable<Element> candidates) {
-    return filter(parent.apply(candidates));
+  Iterable<CandidateType> findInCandidates(Iterable<CandidateType> candidates) {
+    return filter(parent.findInCandidates(candidates));
   }
 
   @override
-  Iterable<Element> get allCandidates => parent.allCandidates;
+  Iterable<CandidateType> get allCandidates => parent.allCandidates;
 }
 
-class _FirstFinder extends ChainedFinder {
-  _FirstFinder(super.parent);
+/// Applies additional filtering against a [parent] [Finder].
```

Comment:
Maybe add to the docs that this one is specifically about `Widget`s

Reply:
Done.